src/linalg.rs

```diff
@@ -0,0 +1,696 @@
+        (Some(Ord::P(p)), Some(axes)) => norm_p_device(array, p, axes, keep_dims, stream),
+    }
+}
+
+/// The QR factorization of the input matrix. Returns an error if the input is not valid.
+///
+/// This function supports arrays with at least 2 dimensions. The matrices which are factorized are
+/// assumed to be in the last two dimensions of the input.
+///
+/// Evaluation on the GPU is not yet implemented.
+///
+/// # Params
+///
+/// - `array`: input array
+///
+/// # Example
+///
+/// ```rust
+/// use mlx_rs::{prelude::*, linalg::*};
+///
+/// let a = Array::from_slice(&[2.0f32, 3.0, 1.0, 2.0], &[2, 2]);
+///
+/// let (q, r) = qr_device(&a, StreamOrDevice::cpu()).unwrap();
+///
+/// let q_expected = Array::from_slice(&[-0.894427, -0.447214, -0.447214, 0.894427], &[2, 2]);
+/// let r_expected = Array::from_slice(&[-2.23607, -3.57771, 0.0, 0.447214], &[2, 2]);
+///
+/// assert!(q.all_close(&q_expected, None, None, None).unwrap().item::<bool>());
+/// assert!(r.all_close(&r_expected, None, None, None).unwrap().item::<bool>());
+/// ```
+#[default_device]
+pub fn qr_device(a: &Array, stream: impl AsRef<Stream>) -> Result<(Array, Array), Exception> {
+    unsafe {
+        let c_vec = try_catch_c_ptr_expr! {
+            mlx_sys::mlx_linalg_qr(a.as_ptr(), stream.as_ref().as_ptr())
+        };
+
+        let v = VectorArray::from_ptr(c_vec);
+
+        let vals: SmallVec<[Array; 2]> = v.into_values();
+        let mut iter = vals.into_iter();
+        let q = iter.next().unwrap();
+        let r = iter.next().unwrap();
+
+        Ok((q, r))
+    }
+}
+
+/// The Singular Value Decomposition (SVD) of the input matrix.
+///
+/// This function supports arrays with at least 2 dimensions. When the input has more than two
+/// dimensions, the function iterates over all indices of the first a.ndim - 2 dimensions and for
+/// each combination SVD is applied to the last two indices.
+///
+/// Evaluation on the GPU is not yet implemented.
+///
+/// # Params
+///
+/// - `array`: input array
+///
+/// # Safety
+///
+/// This is unsafe because it does not check if the arguments are valid.
+///
+/// # Example
+///
+/// ```rust
+/// use mlx_rs::{prelude::*, linalg::*};
+///
+/// let a = Array::from_slice(&[1.0f32, 2.0, 3.0, 4.0], &[2, 2]);
+/// let (u, s, vt) = unsafe { svd_device_unchecked(&a, StreamOrDevice::cpu()) };
+/// let u_expected = Array::from_slice(&[-0.404554, 0.914514, -0.914514, -0.404554], &[2, 2]);
+/// let s_expected = Array::from_slice(&[5.46499, 0.365966], &[2]);
+/// let vt_expected = Array::from_slice(&[-0.576048, -0.817416, -0.817415, 0.576048], &[2, 2]);
+/// assert!(u.all_close(&u_expected, None, None, None).unwrap().item::<bool>());
+/// assert!(s.all_close(&s_expected, None, None, None).unwrap().item::<bool>());
+/// assert!(vt.all_close(&vt_expected, None, None, None).unwrap().item::<bool>());
+/// ```
+#[default_device]
+pub unsafe fn svd_device_unchecked(
```

Comment:
`unchecked_` on purpose?

Reply:
Ahh, forgot to remove this

Reply:
Resolved in [6202a12](https://github.com/oxideai/mlx-rs/pull/51/commits/6202a129a0267d7b45dcfd1b6fe6162934480eed)